### PR TITLE
Transducer Decoding: 2.4x faster decoding with confidence

### PR DESCRIPTION
**Pull request by artbataev**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

- removes `BatchedAlignments` data structure from transducer (RNN-T/TDT) greedy decoding
- preserves all required information (confidence, logits, all steps if required) in `BatchedHyps` data structure
- unbiased logits (without LM scores) are used for confidence estimation, and stored directly if required (sum of logits without log-softmax does not make sense)
- confidence estimation no longer requires to preserve all decoding steps (by default blank steps are excluded), this results in huge memory saving and speedup
- WIND RNN-T decoding (making RNN-T as fast as TDT) is incompatible with saving results at each step. This PR unblocks https://github.com/NVIDIA-NeMo/NeMo/pull/14451
- add `confidence` flag to `examples/asr/asr_chunked_inference/rnnt/speech_to_text_streaming_infer_rnnt.py` to output word confidence

Decoding LibriSpeech test-other, bs=64 (full model, not only decoder!), `nvidia/parakeet-rnnt-1.1b`:

| Approach | Confidence | Alignments*    | RTFx | WER   | Comment                        |
|----------|------------|----------------|------|-------|--------------------------------|
| `main`   | no         | no             | 1734 | 2.68% | Base decoding                  |
| `main`   | yes        | yes (required) | 683  | 2.68% | 2.5x slower (!)                |
| new      | no         | no             | 1730 | 2.68% | Base decoding - same speed/WER |
| new      | yes        | no             | 1655 | 2.68% | 5% overhead for confidence     |

*Alignment = data structure in `Hypothesis` object that stores logits for all decoding steps. 

**Collection**: [ASR]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

```python
import nemo.collections.asr as nemo_asr
from omegaconf import open_dict
asr_model = nemo_asr.models.ASRModel.from_pretrained(model_name="nvidia/parakeet-rnnt-1.1b")
with open_dict(asr_model.cfg.decoding):
    asr_model.cfg.decoding.confidence_cfg = dict(preserve_frame_confidence=True, preserve_token_confidence=True, preserve_word_confidence=True)
asr_model.change_decoding_strategy(asr_model.cfg.decoding)
manifest = "<manifest.jsonl>"
hyps = asr_model.transcribe(manifest, return_hypotheses=True)
print(hyps[0].word_confidence)
print(hyps[0].token_confidence)
```

```shell
# word confidence in streaming decoding
python examples/asr/asr_chunked_inference/rnnt/speech_to_text_streaming_infer_rnnt.py \
    <model_and_data_params> \
    confidence=true
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/parts/submodules/transducer_decoding/label_looping_base.py [hunk 1]
```diff
@@ -138,6 +140,64 @@ def disable_cuda_graphs(self) -> bool:
         self.reset_cuda_graphs_state()
         return True
 
+    # fusion models-related methods
```
Comment (lilithgrigoryan) [inline nemo/collections/asr/parts/submodules/transducer_decoding/label_looping_base.py:152]: Are these changes related to the PR? Seems like this is more related to per-stream biasing PR

**Review by lilithgrigoryan — CHANGES_REQUESTED** [review]

**Review by lilithgrigoryan — DISMISSED** [review]
Thanks @artbataev. LGTM!

**Review by lilithgrigoryan — APPROVED** [review]
